src/main/java/com/google/googleidentity/oauth2/jwt/JwtAssertionRequestHandler.java

```diff
@@ -0,0 +1,229 @@
+import java.util.logging.Logger;
+import javax.servlet.http.HttpServletResponse;
+import net.minidev.json.JSONObject;
+import org.apache.http.HttpStatus;
+
+/** Processor for dealing JWT Assertion */
+@Singleton
+public class JwtAssertionRequestHandler implements RequestHandler {
+
+  private static final String keyUrl = "https://www.googleapis.com/oauth2/v3/certs";
+  private static final String aud =
+      "475640046628-i42g5qfbcp58e3nijqiedomvhe7hb3sn.apps.googleusercontent.com";
+  private final OAuth2TokenService oauth2TokenService;
+  private final UserDetailsService userDetailsService;
+
+  private final ClientDetailsService clientDetailsService;
+
+  private final Logger log = Logger.getLogger("JwtAssertionTokenProcessor");
+
+  @Inject
+  public JwtAssertionRequestHandler(
+      OAuth2TokenService oauth2TokenService,
+      UserDetailsService userDetailsService,
+      ClientDetailsService clientDetailsService) {
+    this.oauth2TokenService = oauth2TokenService;
+    this.userDetailsService = userDetailsService;
+    this.clientDetailsService = clientDetailsService;
+  }
+
+  @Override
+  public void handle(HttpServletResponse response, OAuth2Request oauth2Request)
+      throws IOException, OAuth2Exception {
+    List<String> info =
+        verifyAndGetInfoFromJwt(
+            oauth2Request.getRequestBody().getAssertion(), new JwtSigningKeyResolver(keyUrl));
+    String email = info.get(0);
+    String googleAccountId = info.get(1);
+    switch (oauth2Request.getRequestBody().getIntent()) {
+      case CHECK:
+        handleCheckAssertion(response, email, googleAccountId);
+        break;
+      case GET:
+        handleGetAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      case CREATE:
+        handleCreateAssertion(
+            response, email, googleAccountId, oauth2Request.getRequestBody().getScopesList());
+        break;
+      default:
+        throw new IllegalStateException();
+    }
+  }
+
+  public List<String> verifyAndGetInfoFromJwt(String assertion, SigningKeyResolverAdapter keys)
+      throws OAuth2Exception {
+    Jws<Claims> jws;
+
+    String email = null;
+    String googleAccountId = null;
+    try {
+      jws = Jwts.parserBuilder().setSigningKeyResolver(keys).build().parseClaimsJws(assertion);
+      if (!jws.getBody().getIssuer().equals("https://accounts.google.com")) {
+        throw new InvalidRequestException(ErrorCode.INVALID_JWT_ISS);
+      }
+      if (!jws.getBody().getAudience().equals(aud)) {
+        throw new InvalidRequestException(ErrorCode.WRONG_JWT_AUD);
+      }
+      email = jws.getBody().get("email", String.class);
+      googleAccountId = jws.getBody().get("sub", Long.class).toString();
+    } catch (JwtException ex) {
+      log.log(Level.INFO, "JWT Decode ERROR!", ex);
+      throw new InvalidRequestException(ErrorCode.INVALID_JWT);
+    }
+    return ImmutableList.of(email, googleAccountId);
```

Comment:
You can use an inner class with [@AutoValue](https://github.com/google/auto/tree/master/value) or at least Pair<String, String>